services/grpc/grpc.js

```diff
@@ -1,278 +1,211 @@
+import lightningSubscriptions from './lightning.subscriptions'
 
-const WALLET_STATE_LOCKED = 'WALLET_STATE_LOCKED'
-const WALLET_STATE_ACTIVE = 'WALLET_STATE_ACTIVE'
+const GRPC_WALLET_UNLOCKER_SERVICE_ACTIVE = 'GRPC_WALLET_UNLOCKER_SERVICE_ACTIVE'
+const GRPC_LIGHTNING_SERVICE_ACTIVE = 'GRPC_LIGHTNING_SERVICE_ACTIVE'
 
 /**
- * Wrapper and controller for multiple GrpcService instances.
+ * LND gRPC wrapper.
  * @extends EventEmitter
  */
-class Grpc extends EventEmitter {
-  constructor() {
-    super()
-    this.fsm = new StateMachine({
-      init: 'pending',
-      transitions: [
-        { name: 'init', from: 'pending', to: 'ready' },
-        { name: 'activateWalletUnlocker', from: ['ready', 'active'], to: 'locked' },
-        { name: 'activateLightning', from: ['ready', 'locked'], to: 'active' },
-        { name: 'disconnect', from: ['locked', 'active'], to: 'ready' },
-      ],
-      methods: {
-        onBeforeActivateWalletUnlocker: this.onBeforeActivateWalletUnlocker.bind(this),
-        onBeforeActivateLightning: this.onBeforeActivateLightning.bind(this),
-        onBeforeDisconnect: this.onBeforeDisconnect.bind(this),
-        onAfterDisconnect: this.onAfterDisconnect.bind(this),
-        onLeaveActive: this.onLeaveActive.bind(this),
-      },
-    })
-    this.supportedServices = [WalletUnlocker, Lightning]
-    this.services = {}
-    this.fsm.init()
-  }
+class ZapGrpc extends EventEmitter {
+  init(options) {
+    this.options = options
+    this.subscriptions = []
 
-  /**
-   * Initialize the service.
-   * @param {LndConfig} lndConfig configuration.
-   */
-  init(lndConfig) {
-    grpcLog.info(`Initializing lnd gRPC with config: %o`, lndConfig)
-    this.lndConfig = lndConfig
+    // Create a new grpc instance using settings from init options.
+    const grpcOptions = this._getConnectionSettings()
+    this.grpc = new LndGrpc(grpcOptions)
 
-    // Instantiate individual services.
-    this.supportedServices.forEach(Service => {
-      const instance = new Service(this.lndConfig)
-      this.services[instance.serviceName] = instance
-    })
+    // Set up service accessors.
+    this.services = {}
+    Object.keys(this.grpc.services).reduce((accumulator, currentValue) => {
+      accumulator[currentValue] = this.grpc.services[currentValue]
+      return accumulator
+    }, this.services)
 
-    // Activate the lightning interfae as soon as the WalletUnlocker has been unlocked.
-    this.services.WalletUnlocker.on('UNLOCK_WALLET_SUCCESS', async () => {
-      this.activateLightning()
-    })
-  }
+    const { Lightning } = this.services
 
-  // ------------------------------------
-  // FSM Proxies
-  // ------------------------------------
+    // Inject helper methods.
+    Object.assign(Lightning, lightningMethods)
+    Object.assign(Lightning, lightningSubscriptions)
 
-  is(...args) {
-    return this.fsm.is(args)
-  }
+    // Setup gRPC event forwarders.
+    this.grpc.on('locked', () => {
```

Comment:
We don't, but it seemed to me that there is no need to. Why we want to stop forwarding events? If the service is disconnected, it will stop emitting events.

Reply:
As long as we are sure `init` is only called once

Reply:
I've added a guard at the top of the `init` method to prevent multiple calls to `init`.

Reply:
Actually, adding the guard highlighted the fact that cleanup is in-fact needed. Given that this class is used as a singleton it means that it gets reused from one node connection to another. To better cater for this I have merged the `init` and `connect` methods into a single `connect` method with a guard to prevent callers from connecting when it's already connected, and in addition I have added relevant cleanup and rest code into the disconnect.

So users of the class must call `disconnect` to cleanup before calling `connect` again (with potentially new parameters)